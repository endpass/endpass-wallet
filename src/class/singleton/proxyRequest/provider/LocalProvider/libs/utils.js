import regPath from 'path-to-regexp';

const utils = {
  getRouteKey(routesRegexp, routePath) {
    return Object.keys(routesRegexp).find(key => {
      const re = routesRegexp[key];
      return re.test(routePath);
    });
  },
  getPath(serverUrl, url) {
    const levels = url.split(serverUrl);
    const keysPath = levels.pop() || '';
    const ret = keysPath[0] === '/' ? keysPath.substring(1) : keysPath;

    return ret;
  },
  getPathArgs(routesRegexp, key, routePath) {
    const keys = regPath.parse(key);
    const re = routesRegexp[key];
    re.lastIndex = 0; // drop exec position

    const values = re.exec(routePath);
    return keys.reduce((map, item, index) => {
      if (item && item.name) {
        map[item.name] = values[index];
      }
      return map;
    }, {});
  },
};

export default utils;
