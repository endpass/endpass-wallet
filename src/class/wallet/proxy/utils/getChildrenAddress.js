export default function(hdWallet, offset, limit) {
  return [...Array(limit)].map((_, i) =>
    hdWallet
      .deriveChild(offset + i)
      .getWallet()
      .getChecksumAddressString(),
  );
}
