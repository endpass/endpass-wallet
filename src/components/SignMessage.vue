<template>
  <div>
    <v-form
      id="signing-message-form"
      data-test="signing-message-form"
      @submit="togglePasswordModal"
    >
      <v-textarea
        v-model="message"
        label="Message"
        placeholder="This is a message that you are signing to prove that you own the address you say you own."
        data-test="message-textarea"
      />
      <v-button
        :disabled="!message"
        class-name="is-primary is-medium"
        data-test="sign-button"
      >Sign message</v-button>
      <v-textarea
        v-if="signedMessage"
        v-model="getSignedMessage"
        label="Signature"
        disabled
        data-test="signed-message-textarea"
      />
    </v-form>
    <password-modal
      v-if="isPasswordModal"
      @confirm="signMessage"
      @close="togglePasswordModal"
    >
      <div class="field">
        <label class="label">Message</label>
        <p>{{ message }}</p>
      </div>
    </password-modal>
  </div>
</template>

<script>
import { mapGetters } from 'vuex';
import PasswordModal from '@/components/modal/PasswordModal';
import modalMixin from '@/mixins/modal';

export default {
  name: 'SignMessage',

  data: () => ({
    message: '',
    signedMessage: null,
  }),

  computed: {
    ...mapGetters('accounts', ['wallet']),

    getSignedMessage() {
      return JSON.stringify(this.signedMessage);
    },
  },

  methods: {
    async signMessage(password) {
      try {
        this.togglePasswordModal();
        this.signedMessage = await this.wallet.sign(this.message, password);
      } catch (error) {
        this.signedMessage = null;
        this.$notify({
          title: 'Error signing message',
          text:
            'An error occurred while signing the message. Please try again.',
          type: 'is-danger',
        });
        console.error(error);
      }
    },
  },

  mixins: [modalMixin],

  components: {
    PasswordModal,
  },
};
</script>
