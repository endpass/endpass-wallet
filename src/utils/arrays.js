export const uniq = arr =>
  arr.reduce((acc, item) => (acc.includes(item) ? acc : acc.concat(item)), []);

export const mapArrayByProp = (arr, prop) =>
  arr.reduce((acc, item) => {
    const target = item[prop];

    if (target) {
      return Object.assign(acc, {
        [target]: item,
      });
    }

    return acc;
  }, {});

export default {
  uniq,
  mapArrayByProp,
};
