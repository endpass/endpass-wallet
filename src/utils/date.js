import dayjs from 'dayjs';
import relativeTime from 'dayjs/plugin/relativeTime';

/**
 * Setting up dayjs globally
 */
dayjs.extend(relativeTime);

export const formateDate = date => dayjs(date).format('YYYY-MM-DD H:mm');

export const fromNow = date => dayjs(date).fromNow();

export default {
  formateDate,
  fromNow,
};
