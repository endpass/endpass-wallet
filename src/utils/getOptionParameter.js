export default function getOptionParameter(item, value) {
  return item instanceof Object ? item[value] : item;
}
