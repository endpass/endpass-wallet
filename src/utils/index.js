import store from '../store';

export { default as http } from './http';

export const kebabToCamel = input =>
  input.replace(/-([a-z])/g, g => g[1].toUpperCase());

export const getInitializedValueFromStore = (module, field) => {
  return new Promise(resolve => {
    const value = module[field];

    if (value === null) {
      const unwatch = store.watch(
        () => module[field],
        newValue => {
          unwatch();
          resolve(newValue);
        },
      );
    } else {
      resolve(value);
    }
  });
};
