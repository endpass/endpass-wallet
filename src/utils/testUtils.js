import { kebabCase } from 'lodash';
import { getFrom } from '@/utils/objects';

function convertDirectivesToAttrs(directives) {
  const attrs = {};

  if (directives) {
    directives.forEach(({ rawName, expression }) => {
      attrs[rawName] = expression;
    });
  }

  return attrs;
}

function convertListenersToAttrs(listeners) {
  const attrs = {};

  if (listeners) {
    Object.keys(listeners).forEach(key => {
      const handlers = [].concat(listeners[key]);
      const handler = handlers.find(h => {
        const name = getFrom(h, 'fns.name', 'name') || '';

        return name.includes('bound ');
      });

      if (handler) {
        attrs[`v-on:${key}`] = (handler.fns || handler).name;
      }
    });
  }

  return attrs;
}

function convertModelToAttrs(model) {
  return model ? { 'v-model': model.expression } : {};
}

function convertSlotToAttrs(slot) {
  return slot ? { slot } : {};
}

function generateElement(createElement, vnode) {
  return vnode.componentOptions
    ? generateElementFromComponent(createElement, vnode)
    : generateElementFromHTML(createElement, vnode);
}

function getChildrenElements(createElement, children) {
  return children
    ? children.map(child => generateElement(createElement, child))
    : undefined;
}

function generateElementFromComponent(
  createElement,
  { data, componentOptions },
) {
  const { attrs, directives, model } = data;
  const { tag, listeners, children } = componentOptions;
  const componentData = {
    ...data,
    attrs: {
      ...attrs,
      ...convertDirectivesToAttrs(directives),
      ...convertListenersToAttrs(listeners),
      ...convertModelToAttrs(model),
      class: data.staticClass,
    },
    hook: undefined,
    staticClass: undefined,
  };

  if (listeners) {
    componentData.on = Object.assign({}, listeners);
  }

  return createElement(
    tag,
    componentData,
    getChildrenElements(createElement, children),
  );
}

function generateElementFromHTML(createElement, vnode) {
  if (!vnode.data) {
    return vnode;
  }

  const { tag, data, children } = vnode;
  const { directives, on, model, attrs = {}, slot } = data;

  return createElement(
    tag,
    {
      ...data,
      attrs: {
        ...attrs,
        ...convertDirectivesToAttrs(directives),
        ...convertListenersToAttrs(on),
        ...convertModelToAttrs(model),
        ...convertSlotToAttrs(slot),
      },
    },
    getChildrenElements(createElement, children),
  );
}

export function generateStubs(Component) {
  return Object.values(Component.components).reduce((stubs, stubComponent) => {
    let elementName;

    if (stubComponent.name) {
      elementName = kebabCase(stubComponent.name);
    }

    return Object.assign(stubs, {
      [elementName]: {
        render(createElement) {
          return generateElementFromComponent(createElement, this.$vnode);
        },
      },
    });
  }, {});
}

export default {
  generateStubs,
};
