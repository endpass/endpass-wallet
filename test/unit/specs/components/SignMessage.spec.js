import Vuex from 'vuex';
import { shallow, createLocalVue } from '@vue/test-utils';
import Notifications from 'vue-notification';
import SignMessage from '@/components/SignMessage';
import { generateStubs } from '@/utils/testUtils';

const localVue = createLocalVue();

localVue.use(Vuex);
localVue.use(Notifications);

describe('SignMessage', () => {
  const signedMessage = {};
  const sign = jest.fn().mockResolvedValue(signedMessage);
  let wrapper;
  let store;

  beforeEach(() => {
    store = new Vuex.Store({
      modules: {
        accounts: {
          namespaced: true,
          getters: {
            wallet: () => ({
              getPrivateKeyString: jest.fn(() => 'private key string'),
              sign,
            }),
          },
        },
      },
      state: {
        web3: {},
      },
    });

    wrapper = shallow(SignMessage, {
      localVue,
      store,
      stubs: generateStubs(SignMessage),
    });
  });

  describe('render', () => {
    it('should render the initial state of the component', () => {
      expect(wrapper.element).toMatchSnapshot();
    });

    it('should enable "Sign message" button', () => {
      wrapper.setData({
        message: 'message',
      });

      expect(wrapper.element).toMatchSnapshot();
    });

    it('should show signed message', () => {
      const signedMessage = { message: 'message' };

      wrapper.setData({ signedMessage });

      expect(wrapper.element).toMatchSnapshot();
    });

    it('should show modal window for password confirmation', () => {
      wrapper.setData({
        message: 'message',
        isPasswordModal: true,
      });

      expect(wrapper.element).toMatchSnapshot();
    });
  });

  describe('methods', () => {
    describe('signMessage', () => {
      const password = 'password';

      beforeEach(() => {
        jest.spyOn(wrapper.vm, '$notify');
      });

      it('should sign message', async () => {
        expect.assertions(2);

        const { vm } = wrapper;

        await vm.signMessage(password);

        expect(vm.signedMessage).toEqual(signedMessage);
        expect(vm.$notify).not.toHaveBeenCalled();
      });

      it('should not sign message', async () => {
        expect.assertions(3);

        const { vm } = wrapper;

        sign.mockRejectedValueOnce(new Error());

        global.console.error = jest.fn();

        await vm.signMessage(password);

        expect(vm.signedMessage).toBeNull();
        expect(vm.$notify).toHaveBeenCalledTimes(1);
        expect(vm.$notify).toHaveBeenCalledWith({
          title: 'Error signing message',
          text:
            'An error occurred while signing the message. Please try again.',
          type: 'is-danger',
        });
      });
    });
  });
});
