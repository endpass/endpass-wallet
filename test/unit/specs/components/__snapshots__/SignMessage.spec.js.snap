// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SignMessage render should enable "Sign message" button 1`] = `
<div>
  <v-form
    data-test="signing-message-form"
    id="signing-message-form"
    v-on:submit="bound togglePasswordModal"
  >
    <v-textarea
      data-test="message-textarea"
      label="Message"
      placeholder="This is a message that you are signing to prove that you own the address you say you own."
      v-model="message"
    />
     
    <v-button
      class-name="is-primary is-medium"
      data-test="sign-button"
    >
      Sign message
    </v-button>
     
    <!---->
  </v-form>
   
  <!---->
</div>
`;

exports[`SignMessage render should render the initial state of the component 1`] = `
<div>
  <v-form
    data-test="signing-message-form"
    id="signing-message-form"
    v-on:submit="bound togglePasswordModal"
  >
    <v-textarea
      data-test="message-textarea"
      label="Message"
      placeholder="This is a message that you are signing to prove that you own the address you say you own."
      v-model="message"
    />
     
    <v-button
      class-name="is-primary is-medium"
      data-test="sign-button"
      disabled="true"
    >
      Sign message
    </v-button>
     
    <!---->
  </v-form>
   
  <!---->
</div>
`;

exports[`SignMessage render should show modal window for password confirmation 1`] = `
<div>
  <v-form
    data-test="signing-message-form"
    id="signing-message-form"
    v-on:submit="bound togglePasswordModal"
  >
    <v-textarea
      data-test="message-textarea"
      label="Message"
      placeholder="This is a message that you are signing to prove that you own the address you say you own."
      v-model="message"
    />
     
    <v-button
      class-name="is-primary is-medium"
      data-test="sign-button"
    >
      Sign message
    </v-button>
     
    <!---->
  </v-form>
   
  <password-modal
    v-on:close="bound togglePasswordModal"
    v-on:confirm="bound signMessage"
  >
    <div
      class="field"
    >
      <label
        class="label"
      >
        Message
      </label>
       
      <p>
        message
      </p>
    </div>
  </password-modal>
</div>
`;

exports[`SignMessage render should show signed message 1`] = `
<div>
  <v-form
    data-test="signing-message-form"
    id="signing-message-form"
    v-on:submit="bound togglePasswordModal"
  >
    <v-textarea
      data-test="message-textarea"
      label="Message"
      placeholder="This is a message that you are signing to prove that you own the address you say you own."
      v-model="message"
    />
     
    <v-button
      class-name="is-primary is-medium"
      data-test="sign-button"
      disabled="true"
    >
      Sign message
    </v-button>
     
    <v-textarea
      data-test="signed-message-textarea"
      disabled=""
      label="Signature"
      v-model="getSignedMessage"
    />
  </v-form>
   
  <!---->
</div>
`;
